nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java

```diff
@@ -0,0 +1,235 @@
+        final JsonBuilderFactory factory = Json.createBuilderFactory(config);
+        final JsonObjectBuilder builder = factory.createObjectBuilder();
+
+        final DateFormat df = new SimpleDateFormat(TIMESTAMP_FORMAT);
+        df.setTimeZone(TimeZone.getTimeZone("Z"));
+
+        final long start = System.nanoTime();
+
+        // Create a JSON array of all the events in the current batch
+        final JsonArrayBuilder arrayBuilder = factory.createArrayBuilder();
+        for (final Bulletin bulletin : bulletins) {
+            if(bulletin.getId() > lastSentBulletinId) {
+                arrayBuilder.add(serialize(factory, builder, bulletin, df, platform, nodeId));
+            }
+        }
+        final JsonArray jsonArray = arrayBuilder.build();
+
+        // Send the JSON document for the current batch
+        try {
+            final Transaction transaction = getClient().createTransaction(TransferDirection.SEND);
+            if (transaction == null) {
+                getLogger().debug("All destination nodes are penalized; will attempt to send data later");
+                return;
+            }
+
+            final Map<String, String> attributes = new HashMap<>();
+            final String transactionId = UUID.randomUUID().toString();
+            attributes.put("reporting.task.transaction.id", transactionId);
+
+            final byte[] data = jsonArray.toString().getBytes(StandardCharsets.UTF_8);
+            transaction.send(data, attributes);
+            transaction.confirm();
+            transaction.complete();
+
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+            getLogger().info("Successfully sent {} Bulletins to destination in {} ms; Transaction ID = {}; First Event ID = {}",
+                    new Object[]{bulletins.size(), transferMillis, transactionId, bulletins.get(0).getId()});
+        } catch (final IOException e) {
+            throw new ProcessException("Failed to send Bulletins to destination due to IOException:" + e.getMessage(), e);
+        }
+
+        // Store the id of the last event so we know where we left off
+        try {
+            StateManager stateManager = context.getStateManager();
+            Map<String, String> newMapOfState = new HashMap<>();
+            newMapOfState.put(LAST_EVENT_ID_KEY, String.valueOf(currMaxId));
+            stateManager.setState(newMapOfState, Scope.LOCAL);
+        } catch (final IOException ioe) {
+            getLogger().error("Failed to update state to {} due to {}; this could result in events being re-sent after a restart.",
+                    new Object[]{currMaxId, ioe.getMessage()}, ioe);
+        }
+
+        lastSentBulletinId = currMaxId;
+    }
+
+    private Long getMaxBulletinId(List<Bulletin> bulletins) {
+        long result = -1L;
+        for (Bulletin bulletin : bulletins) {
+            if(bulletin.getId() > result) {
+                result = bulletin.getId();
```

Comment:
Hmmm, just trying to get the maximum ID from a list of bulletins. I do need to parse the whole list to get the maximum ID (unless I assume there is an order). I updated the PR with a more concise approach. Let me know if I missed something.